src/dolphin/os/OSInit.c

```diff
@@ -70,603 +105,312 @@ void ClearArena(void)
-/* 803432DC 0033FEBC  80 03 00 00 */	lwz r0, 0(r3)
-/* 803432E0 0033FEC0  28 00 00 02 */	cmplwi r0, 2
-/* 803432E4 0033FEC4  41 80 00 08 */	blt lbl_803432EC
-/* 803432E8 0033FEC8  4B FE 73 D5 */	bl EnableMetroTRKInterrupts
-lbl_803432EC:
-/* 803432EC 0033FECC  4B FF FB D1 */	bl ClearArena
-/* 803432F0 0033FED0  48 00 40 89 */	bl OSEnableInterrupts
-lbl_803432F4:
-/* 803432F4 0033FED4  80 01 00 14 */	lwz r0, 0x14(r1)
-/* 803432F8 0033FED8  83 E1 00 0C */	lwz r31, 0xc(r1)
-/* 803432FC 0033FEDC  83 C1 00 08 */	lwz r30, 8(r1)
-/* 80343300 0033FEE0  7C 08 03 A6 */	mtlr r0
-/* 80343304 0033FEE4  38 21 00 10 */	addi r1, r1, 0x10
-/* 80343308 0033FEE8  4E 80 00 20 */	blr
-} // clang-format on
-#pragma pop
 
-#else
+typedef struct BI2Debug {
+    s32 debugMonSize;  // 0x0
+    s32 simMemSize;    // 0x4
+    u32 argOffset;     // 0x8
+    u32 debugFlag;     // 0xC
+    int trackLocation; // 0x10
+    int trackSize;     // 0x14
+    u32 countryCode;   // 0x18
+    u8 unk[8];         // 0x1C
+    u32 padSpec;       // 0x24
+} BI2Debug;
+
+#define OS_BI2_DEBUG_ADDRESS 0x800000F4
+#define OS_DEBUG_ADDRESS_2 0x800030E9
+
+extern vu16 __OSDeviceCode AT_ADDRESS(OS_BASE_CACHED | 0x30E6);
+
+extern volatile struct {
+    u32 x0;
+    u32 intr;
+    u8 unused[0x2C - 0x8];
+    u32 x2C;
+} OS_PI AT_ADDRESS(0xCC003000);
 
 void OSInit(void)
 {
-    NOT_IMPLEMENTED;
-}
-
-#endif
-
-#ifdef MWERKS_GEKKO
+    /*
+    Initializes the Dolphin operating system.
+        - most of the main operations get farmed out to other functions
+        - loading debug info and setting up heap bounds largely happen here
+        - a lot of OS reporting also gets controlled here
+    */
+    // pretty sure this is the min(/max) amount of pointers etc for the stack to
+    // match
+    BI2Debug* DebugInfo;
+    void* debugArenaLo;
+    u32 inputConsoleType;
+
+    // check if we've already done all this or not
+    if (!AreWeInitialized) {
+        AreWeInitialized =
+            true; // flag to make sure we don't have to do this again
+
+        // SYSTEM //
+        __OSStartTime = __OSGetSystemTime();
+        OSDisableInterrupts();
+
+        // DEBUG //
+        // load some DVD stuff
+        BI2DebugFlag = 0; // debug flag from the DVD BI2 header
+        BootInfo = (OSBootInfo*) OS_BASE_CACHED; // set pointer to BootInfo
+
+        __DVDLongFileNameFlag = false; // flag to tell us whether we make it
+                                       // through the debug loading
+
+        // time to grab a bunch of debug info from the DVD
+        // the address for where the BI2 debug info is, is stored at
+        // OS_BI2_DEBUG_ADDRESS
+        DebugInfo = (BI2Debug*) *((u32*) OS_BI2_DEBUG_ADDRESS);
+
+        // if the debug info address exists, grab some debug info
+        if (DebugInfo != NULL) {
+            BI2DebugFlag = &DebugInfo->debugFlag; // debug flag from DVD BI2
+            __PADSpec =
+                (u32) DebugInfo->padSpec; // some other info from DVD BI2
```

Comment:
I won't ask you to change it right now but in the future please place comments above the line they describe, nor beside.